src/gpu/kmeans/kmeans_init.cu

```diff
@@ -0,0 +1,439 @@
+      // centroid (a.k.a y)
+      _distance_pairs.ptr[i*_centroids_dots.rows + idy] =
+          _data_dots.ptr[idx] + _centroids_dots.ptr[idy];
+    }
+  }
+}
+
+// See SelfArgMinOp
+template <typename T>
+__global__ void self_row_argmin_sequential(kParam<int> _res, kParam<T> _val) {
+
+  size_t idx = global_thread_idx();
+  if (idx < _val.rows) {
+    T min = std::numeric_limits<T>::max();
+    int min_idx = -1;
+    for (size_t i = 0; i < _val.cols; ++i) {
+      T value = _val.ptr[idx * _val.cols + i];
+      if (value < min && value != 0) {
+        min = value;
+        min_idx = i;
+      }
+    }
+    _res.ptr[idx] = min_idx;
+  }
+}
+
+template <typename T>
+__global__ void self_row_min_sequential(kParam<T> _res, kParam<T> _val) {
+
+  size_t idx = global_thread_idx();
+  if (idx < _val.rows) {
+    T min = std::numeric_limits<T>::max();
+    for (size_t i = 0; i < _val.cols; ++i) {
+      T value = _val.ptr[idx * _val.cols + i];
+      if (value < min) {
+        min = value;
+      }
+    }
+    min += ESP;
+    _res.ptr[idx] = min;
+  }
+}
+
+}  // namespace kernel
+
+namespace detail {
+
+template <typename T>
+void PairWiseDistanceOp<T>::initialize(KmMatrix<T>& _data_dot,
+                                       KmMatrix<T>& _centroids_dot,
+                                       KmMatrix<T>& _distance_pairs) {
+  data_dot_ = _data_dot;
+  centroids_dot_ = _centroids_dot;
+  distance_pairs_ = _distance_pairs;
+  initialized_ = true;
+}
+
+template <typename T>
+PairWiseDistanceOp<T>::PairWiseDistanceOp(KmMatrix<T>& _data_dot,
+                                          KmMatrix<T>& _centroids_dot,
+                                          KmMatrix<T>& _distance_pairs) :
+    data_dot_(_data_dot), centroids_dot_(_centroids_dot),
+    distance_pairs_(_distance_pairs), initialized_(true) {}
+
+template <typename T>
+KmMatrix<T> PairWiseDistanceOp<T>::operator()(KmMatrix<T>& _data,
+                                              KmMatrix<T>& _centroids) {
+
+  kernel::construct_distance_pairs_kernel<<<
+      dim3(GpuInfo::ins().blocks(32), div_roundup(_centroids.rows(), 16)),
+      dim3(32, 16)>>>(  // FIXME: Tune this.
+          distance_pairs_.k_param(),
+          data_dot_.k_param(),
+          centroids_dot_.k_param());
+
+  safe_cuda(cudaGetLastError());
+
+  cublasHandle_t handle = GpuInfo::ins().cublas_handle();
+
+  T alpha = -2.0;
+  T beta = 1.0;
+
+  Blas::gemm(
+      handle,
+      CUBLAS_OP_T, CUBLAS_OP_N,
+      // n, d, d/k
+      _centroids.rows(), _data.rows(), _data.cols(),
+      &alpha,
+      _centroids.dev_ptr(), _centroids.cols(),
+      _data.dev_ptr(), _data.cols(),
+      &beta,
+      distance_pairs_.dev_ptr(), _centroids.rows());
+
+  return distance_pairs_;
+}
+
+// ArgMin operation that excludes 0. Used when dealing with distance_pairs
+// in recluster where distance between points with itself is calculated,
+// hence the name.
+// FIXME: Maybe generalize it to selection algorithm.
```

Comment:
These two kernels look very similar. Consider generalising them. Less code is better code. Also have a look at this: https://nvlabs.github.io/cub/structcub_1_1_device_segmented_reduce.html

Cub is another really great cuda library.

Reply:
I will look into it. Thanks for the suggestion. Indeed, those extra kernels are quite annoying.